### PR TITLE
Port idp-fixture to testcontainers

### PR DESCRIPTION
**Pull request by breskeby**
This ports idp-fixture to test container and updates downstream tests accordingly.

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -97,6 +97,7 @@ List projects = [
   'test:fixtures:minio-fixture',
   'test:fixtures:old-elasticsearch',
   'test:fixtures:s3-fixture',
+  'test:fixtures:testcontainer-utils',
```
Comment (breskeby) [inline settings.gradle:100]: have a project where we keep general test container utils. e.g. common ThreadLeakFilter or base classes

File: test/test-clusters/src/main/java/org/elasticsearch/test/cluster/ConfigFileProvider.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ */
+
+package org.elasticsearch.test.cluster;
+
+import org.elasticsearch.test.cluster.local.LocalClusterSpec;
+import org.elasticsearch.test.cluster.util.resource.Resource;
+
+import java.util.Map;
+
+public interface ConfigFileProvider {
```
Comment (breskeby) [inline test/test-clusters/src/main/java/org/elasticsearch/test/cluster/ConfigFileProvider.java:16]: Allow lazy ConfigFile configuration

File: x-pack/qa/openldap-tests/src/test/java/org/elasticsearch/xpack/security/authc/ldap/SearchGroupsResolverTests.java [hunk 3]
```diff
@@ -202,7 +212,7 @@ public void testReadBinaryUserAttribute() throws Exception {
 
     @Override
     protected String ldapUrl() {
```
Comment (breskeby) [inline x-pack/qa/openldap-tests/src/test/java/org/elasticsearch/xpack/security/authc/ldap/SearchGroupsResolverTests.java:214]: This is currently not used anymore because of the way `GroupsResolverTestCase` is setup. We need to cleanup GroupsResolverTestCase to not rely on class path resources that are outside of a jar. it shouldn't matter. But this rework is out of scope of this PR.

File: x-pack/qa/openldap-tests/src/test/java/org/elasticsearch/xpack/security/authc/ldap/SearchGroupsResolverTests.java [hunk 4]
```diff
@@ -217,6 +227,12 @@ protected String bindPassword() {
 
     @Override
     protected String trustPath() {
```
Comment (breskeby) [inline x-pack/qa/openldap-tests/src/test/java/org/elasticsearch/xpack/security/authc/ldap/SearchGroupsResolverTests.java:229]: This is currently not used anymore because of the way GroupsResolverTestCase is setup. We need to cleanup GroupsResolverTestCase to not rely on class path resources that are outside of a jar. it shouldn't matter. But this rework is out of scope of this PR.

File: test/fixtures/testcontainer-utils/src/main/java/org/elasticsearch/test/fixtures/testcontainers/TestContainersThreadFilter.java [hunk 5]
```diff
@@ -17,6 +17,8 @@
 public class TestContainersThreadFilter implements ThreadFilter {
     @Override
     public boolean reject(Thread t) {
-        return t.getName().startsWith("testcontainers-") || t.getName().startsWith("ducttape");
+        return t.getName().startsWith("testcontainers-")
+            || t.getName().startsWith("ducttape")
+            || t.getName().startsWith("ForkJoinPool.commonPool-worker-1");
```
Comment (breskeby) [inline test/fixtures/testcontainer-utils/src/main/java/org/elasticsearch/test/fixtures/testcontainers/TestContainersThreadFilter.java:22]: Those are coming from test container internals using `Stream#parallel`. I currently discussing this issue with the testcontainer team in their community slack channel.

File: x-pack/qa/saml-idp-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/saml/SamlAuthenticationIT.java [hunk 6]
```diff
@@ -131,12 +148,24 @@ public class SamlAuthenticationIT extends ESRestTestCase {
         .setting("xpack.security.authc.realms.native.native.order", "4")
         .setting("xpack.ml.enabled", "false")
         .setting("logger.org.elasticsearch.xpack.security", "TRACE")
-        .configFile("sp-signing.key", Resource.fromClasspath("sp-signing.key"))
-        .configFile("idp-metadata.xml", Resource.fromClasspath("idp-metadata.xml"))
-        .configFile("sp-signing.crt", Resource.fromClasspath("sp-signing.crt"))
+        .configFile("sp-signing.key", Resource.fromClasspath("/idp/shibboleth-idp/credentials/sp-signing.key"))
+        .configFile("idp-metadata.xml", () -> Resource.fromString(calculateIdpMetaData()))
```
Comment (mark-vieira) [inline x-pack/qa/saml-idp-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/saml/SamlAuthenticationIT.java:148]: I believe we already have a mechanism for lazily evaluated resources that avoids all the extra complexity of adding a `Supplier<Resource>` method to the builder API. This could be simplified to `Resource.fromString(() -> calculateIdpMetaData()`.

Comment (breskeby) [reply x-pack/qa/saml-idp-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/saml/SamlAuthenticationIT.java]: neat!. missed that one. Updated the PR accordingly

File: x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/OidcProviderTestContainer.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        );
+        withNetworkAliases("oidc-provider");
+        withNetwork(network);
+        addExposedPort(PORT);
+    }
+
+    @Override
+    public void start() {
+        super.start();
+        copyFileToContainer(
```
Comment (mark-vieira) [inline x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/OidcProviderTestContainer.java:36]: Is there a reason we don't do this in an entrypoint script in the container?

**Review by mark-vieira — APPROVED** [review]

File: docs/build.gradle [hunk 8]
```diff
@@ -112,8 +112,9 @@ testClusters.matching { it.name == "yamlRestTest"}.configureEach {
 
   requiresFeature 'es.index_mode_feature_flag_registered', Version.fromString("8.0.0")
 
-  extraConfigFile 'op-jwks.json', project(':x-pack:test:idp-fixture').file("oidc/op-jwks.json")
-  extraConfigFile 'idp-docs-metadata.xml', project(':x-pack:test:idp-fixture').file("idp/shibboleth-idp/metadata/idp-docs-metadata.xml")
+  // TODO Rene: clean up this kind of cross project file references
```
[the remark this answers is not in the document]
Comment (breskeby) [reply docs/build.gradle:116]: I haven't looked into this yet

File: x-pack/qa/openldap-tests/build.gradle [hunk 9]
```diff
@@ -1,18 +1,16 @@
 apply plugin: 'elasticsearch.standalone-test'
```
[the remark this answers is not in the document]
Comment (breskeby) [reply x-pack/qa/openldap-tests/build.gradle:1]: its not `elasticsearch.standalone-rest-test` but `elasticsearch.standalone-test`. IMO we should keep it. The others are rest tests. this one isn't

File: x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/HttpProxyTestContainer.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    private static final Integer PORT = 8888;
+
+    public HttpProxyTestContainer(Network network) {
+        super(
+            new ImageFromDockerfile("es-http-proxy-fixture", false).withDockerfileFromBuilder(
+                builder -> builder.from(DOCKER_BASE_IMAGE).copy("oidc/nginx.conf", "/etc/nginx/nginx.conf").build()
+            ).withFileFromClasspath("oidc/nginx.conf", "/oidc/nginx.conf")
+        );
+        waitingFor(Wait.forHttp("/"));
+        withLogConsumer(new Slf4jLogConsumer(logger()));
```
[the remark this answers is not in the document]
Comment (breskeby) [reply x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/HttpProxyTestContainer.java:30]: I added this one for debugging actually. I'll add it to the other containers too.

File: x-pack/test/idp-fixture/build.gradle [hunk 11]
```diff
@@ -2,8 +2,36 @@ import org.elasticsearch.gradle.VersionProperties
 import org.elasticsearch.gradle.Architecture
 import static org.elasticsearch.gradle.internal.distribution.InternalElasticsearchDistributionTypes.DOCKER;
 
+apply plugin: 'java'
+apply plugin: 'elasticsearch.java'
+apply plugin: 'elasticsearch.cache-test-fixtures'
 apply plugin: 'elasticsearch.test.fixtures'
```
[the remark this answers is not in the document]
Comment (breskeby) [reply x-pack/test/idp-fixture/build.gradle:8]: removed

File: x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/HttpProxyTestContainer.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        addExposedPort(PORT);
+        withNetwork(network);
+
+    }
+
+    @Override
+    public void cache() {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/HttpProxyTestContainer.java:37]: done

File: x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/IdpTestContainer.java [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+import java.io.IOException;
+import java.nio.file.Path;
+
+import static org.elasticsearch.test.fixtures.ResourceUtils.copyResourceToFile;
+
+public final class IdpTestContainer extends DockerEnvironmentAwareTestContainer implements TestRule, CacheableTestFixture {
+
+    public static final String DOCKER_BASE_IMAGE = "openjdk:11.0.16-jre";
+
+    private final TemporaryFolder temporaryFolder = new TemporaryFolder();
```
[the remark this answers is not in the document]
Comment (breskeby) [reply x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/IdpTestContainer.java:26]: wired that cleanup up

File: x-pack/test/idp-fixture/build.gradle [hunk 14]
```diff
@@ -2,8 +2,36 @@ import org.elasticsearch.gradle.VersionProperties
 import org.elasticsearch.gradle.Architecture
 import static org.elasticsearch.gradle.internal.distribution.InternalElasticsearchDistributionTypes.DOCKER;
 
+apply plugin: 'java'
```
[the remark this answers is not in the document]
Comment (breskeby) [reply x-pack/test/idp-fixture/build.gradle:5]: removed

File: test/fixtures/minio-fixture/src/main/java/org/elasticsearch/test/fixtures/minio/MinioTestContainer.java [hunk 15]
```diff
@@ -25,7 +25,7 @@ public MinioTestContainer() {
 
     public MinioTestContainer(boolean enabled) {
         super(
-            new ImageFromDockerfile().withDockerfileFromBuilder(
+            new ImageFromDockerfile("es-minio-testfixture", false).withDockerfileFromBuilder(
```
[the remark this answers is not in the document]
Comment (breskeby) [reply test/fixtures/minio-fixture/src/main/java/org/elasticsearch/test/fixtures/minio/MinioTestContainer.java:28]: My initial idea was to reuse containers. Removed that

File: x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/OidcProviderTestContainer.java (hunk 7 above)
Comment (breskeby) [reply x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/OidcProviderTestContainer.java:36]: As I understand it I would need access to the mapped port at that point. is that supported by an ENTRYPOINT to have access to that. I think it doesn't work how testcontainers are wired up.

File: x-pack/qa/openldap-tests/build.gradle (hunk 9 above)
[the remark this answers is not in the document]
Comment (mark-vieira) [reply x-pack/qa/openldap-tests/build.gradle:1]: Got it. The question remains, what's the point of this plugin? Why don't we just use the `elasticsearch.java` plugin for these projects?

File: x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/HttpProxyTestContainer.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mark-vieira) [reply x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/HttpProxyTestContainer.java]: If we think it's helpful and won't just create a ton of noise.

File: test/fixtures/minio-fixture/src/main/java/org/elasticsearch/test/fixtures/minio/MinioTestContainer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (mark-vieira) [reply test/fixtures/minio-fixture/src/main/java/org/elasticsearch/test/fixtures/minio/MinioTestContainer.java]: Ok, I think we set this in others as well. Let's make them consistent.

File: x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/OidcProviderTestContainer.java (hunk 7 above)
Comment (mark-vieira) [reply x-pack/test/idp-fixture/src/main/java/org/elasticsearch/test/fixtures/idp/OidcProviderTestContainer.java:41]: Hmm, yeah, that makes sense. I thought maybe this would be exposed in the container as an environment variable but seems like no such mechanism exists.

File: test/fixtures/minio-fixture/src/main/java/org/elasticsearch/test/fixtures/minio/MinioTestContainer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (breskeby) [reply test/fixtures/minio-fixture/src/main/java/org/elasticsearch/test/fixtures/minio/MinioTestContainer.java]: Cleaned that up

File: x-pack/qa/openldap-tests/build.gradle (hunk 9 above)
[the remark this answers is not in the document]
Comment (breskeby) [reply x-pack/qa/openldap-tests/build.gradle:1]: the main thing this does is that we just have one sourceSet instead of the default two. I doubt that gives us very much in terms of performance these days tbh

[the remark this answers is not in the document]
Comment (mark-vieira) [reply x-pack/qa/openldap-tests/build.gradle:1]: Right, the `main` source set tasks should end up being skipped due to `NO-SOURCE`. I think that's fine and much better than maintaining a separate plugin that likely uses slightly different conventions.
